File: androidApp/src/main/java/pl/matiz22/cocktailapp/android/home/viewmodels/HomeScreenViewModel.kt
package pl.matiz22.cocktailapp.android.home.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import pl.matiz22.cocktailapp.android.core.states.DataState
import pl.matiz22.cocktails.domain.model.Drinks
import pl.matiz22.cocktails.domain.repository.local.DrinksLocalRepository
import pl.matiz22.core.data.repository.errorMessage

class HomeScreenViewModel(
    private val drinksLocalRepository: DrinksLocalRepository,
) : ViewModel() {
    private val _recentDrinks = MutableStateFlow<DataState<Drinks>>(DataState.Loading)
    val recentDrinks = _recentDrinks.asStateFlow()

    init {
        updateRecentDrinks()
    }

    fun updateRecentDrinks() {
        viewModelScope.launch {
            val databaseResult = drinksLocalRepository.getRecentDrinks()

            if (databaseResult.error != null) {
                _recentDrinks.emit(DataState.Error(databaseResult.error!!.errorMessage))
            } else if (databaseResult.data != null) {
                _recentDrinks.emit(DataState.Success(databaseResult.data!!))
            }
        }
    }
}

File: iosApp/Home/Sources/Home/ViewModels/HomeScreenViewModel.swift
import Core
import Shared

public class HomeScreenViewModel: ObservableObject {
	private let drinksLocalRepository: DrinksLocalRepository = DrinksLocalRepositoryProvider().drinksLocalRepository
	@Published private(set) var recentDrinks: DataState<Drinks> = .loading

	public init() {
		Task {
			await fetchDrinks()
		}
	}

	func fetchDrinks() async {
		await MainActor.run {
			recentDrinks = .loading
		}
		do {
			let drinksData = try await drinksLocalRepository.getRecentDrinks()
			await MainActor.run {
				if let data = drinksData.data {
					recentDrinks = .success(data: data)
				} else if let error = drinksData.error {
					recentDrinks = .error(error: String(reflecting: error))
				}
			}
		} catch {
			await MainActor.run {
				recentDrinks = .error(error: String(localized: "UNKNOWN", bundle: .CoreBundle))
			}
		}
	}
}
